utilities/env_librados.cc

```diff
@@ -0,0 +1,1043 @@
+  #endif
+  }
+
+  // Remove any kind of caching of data from the offset to offset+length
+  // of this file. If the length is 0, then it refers to the end of file.
+  // If the system is not caching the file contents, then this is a noop.
+  rocksdb::Status InvalidateCache(size_t offset, size_t length) {
+  #if 0
+    fs->invalidate_cache(h->file, offset, length);
+  #endif
+    return rocksdb::Status::OK();
+  }
+};
+
+
+// A file abstraction for sequential writing.  The implementation
+// must provide buffering since callers may append small fragments
+// at a time to the file.
+class LibradosWritableFile : public rocksdb::WritableFile {
+    librados::IoCtx _io_ctx;
+    std::string _fid;
+    const int BUFFER_MAX_SIZE = 1 << 12;
+    librados::bufferlist buffer;
+    int buffer_size;
+ public:
+    LibradosWritableFile(librados::IoCtx io_ctx, std::string fid):
+        _io_ctx(io_ctx), _fid(fid), buffer(), buffer_size(0){
+    }
+    ~LibradosWritableFile() {
+        Sync();
+    }
+
+  // Indicates if the class makes use of unbuffered I/O
+  /*bool UseOSBuffer() const {
+    return true;
+    }*/
+
+  // This is needed when you want to allocate
+  // AlignedBuffer for use with file I/O classes
+  // Used for unbuffered file I/O when UseOSBuffer() returns false
+  /*size_t GetRequiredBufferAlignment() const {
+    return c_DefaultPageSize;
+    }*/
+
+  rocksdb::Status Append(const rocksdb::Slice& data) {
+    LOG_DEBUG("%s|%i\n", _fid.c_str(), (int)data.size());
+    #if 0
+    buffer.append(data.data());
+    buffer_size += data.size();
+
+    if (buffer_size > BUFFER_MAX_SIZE) {
+        return Sync();
+    } else {
+        return rocksdb::Status::OK();
+    }
+    #endif
+    librados::bufferlist buf;
+    buf.append(data.data());
+    int r = _io_ctx.append(_fid, buf, data.size());
+    return err_to_status(r);
+  }
+
+  // Positioned write for unbuffered access default forward
+  // to simple append as most of the tests are buffered by default
+  rocksdb::Status PositionedAppend(
+    const rocksdb::Slice& /* data */,
+    uint64_t /* offset */) {
+    return rocksdb::Status::NotSupported();
+  }
+
+  // Truncate is necessary to trim the file to the correct size
+  // before closing. It is not always possible to keep track of the file
+  // size due to whole pages writes. The behavior is undefined if called
+  // with other writes to follow.
+  rocksdb::Status Truncate(uint64_t size) {
+    return rocksdb::Status::OK();
+  }
+
+  rocksdb::Status reset() {
+    std::cout << "reset" << std::endl;
+    return Sync();
+  }
+
+  rocksdb::Status Close() {
+    return Flush();
+  }
+
+  rocksdb::Status Flush() {
+    // TODO: find flush api in librados
+    return Sync();
+  }
+
+  rocksdb::Status Sync() { // sync data
+    // TODO: find sync api in librados
+    if (buffer_size != 0) {
+        int r = _io_ctx.append(_fid, buffer, buffer_size);
+        buffer_size = 0;
+        return err_to_status(r);
```

Comment:
I am making Flush() == Sync() right now. If needed, I will change both API according to your comment.